client_integration_test.go

```diff
@@ -1489,6 +1490,59 @@ func TestClientReadFrom(t *testing.T) {
 	}
 }
 
+// A wrongSizeReader reads zeros and has an arbitrary Size.
+type wrongSizeReader struct{ actual, reported int }
+
+func (r *wrongSizeReader) Read(p []byte) (n int, err error) {
+	if len(p) >= r.actual {
+		n, r.actual = r.actual, 0
+		return n, io.EOF
+	}
+	r.actual -= len(p)
+	return len(p), nil
+}
+
+func (r *wrongSizeReader) Size() int { return r.reported }
+
+// Test File.ReadFrom's handling of a Reader's Size:
+// it should be used as a heuristic for determining concurrency only.
+func TestClientReadFromSizeMismatch(t *testing.T) {
+	const (
+		packetSize = 1024
+		filesize   = 4 * packetSize
+	)
+
+	sftp, cmd := testClient(t, READWRITE, NODELAY, MaxPacketChecked(packetSize), UseConcurrentWrites(true))
+	defer cmd.Wait()
+	defer sftp.Close()
+
+	d, err := ioutil.TempDir("", "sftptest-readfrom-size-mismatch")
+	if err != nil {
+		t.Fatal("cannot create temp dir:", err)
+	}
+	defer os.RemoveAll(d)
+
+	for i, reported := range []int{-1, filesize - 100, filesize, filesize + 100} {
+		r := &wrongSizeReader{filesize, reported}
```

Comment:
(style) Please use explicit field names. Using implicit field assignments is inherently brittle, as changing the order of the fields will changes the order the assignment values have to be in, thus making refactoring harder.

Using explicit field names means not only is it clear exactly which field is getting which value, but also the underlying struct fields can be reordered or even expanded at any time without needing to adjust this code.

Reply:
Done.